### PR TITLE
networkd: Introduce MACsec

### PR DESCRIPTION
**Pull request by ssahani**
This work introduces MACsec to networkd.

Media Access Control Security (MACsec) is an 802.1AE IEEE
industry-standard security technology that provides secure
communication for all traffic on Ethernet links.
MACsec provides point-to-point security on Ethernet links between
directly connected nodes and is capable of identifying and preventing
most security threats, including denial of service, intrusion,
man-in-the-middle, masquerading, passive wiretapping, and playback attacks.

11-macsec.netdev
```
File Edit Options Buffers Tools Help                                                                                                                                                          
[NetDev]
Name=macsec-test
Kind=macsec

[MACsec]
Port=11


[MACsecReceiveChannel]
Port=1234
MACAddress=c6:19:52:8f:e6:a0

[MACsecReceiveChannel]
Port=21
MACAddress=8c:16:45:6c:83:a9

[MACsecTransmitAssociation]
PacketNumber=1024
KeyId=01
Key=81818181818181818181818181818181

[MACsecReceiveAssociation]
Port=1234
MACAddress=c6:19:52:8f:e6:a0
PacketNumber=1
KeyId=00
Key=82828282828282828282828282828282

```

closes https://github.com/systemd/systemd/issues/5754

Please see https://developers.redhat.com/blog/2016/10/14/macsec-a-different-solution-to-encrypt-network-traffic/

https://www.linux.org/docs/man8/ip-macsec.html

### CODE REVIEW

File: src/network/netdev/macsec.c [hunk 1]
```diff
@@ -0,0 +1,915 @@
...
+
+        r = sd_netlink_message_open_container(m, MACSEC_ATTR_SA_CONFIG);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_SA_CONFIG attribute: %m");
+
+        r = sd_netlink_message_append_u8(m, MACSEC_SA_ATTR_AN, a->an);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_SA_ATTR_AN attribute: %m");
+
+        if (a->pn) {
```
Comment (yuwata) [inline src/network/netdev/macsec.c]: `if (a->pn > 0)`

File: src/network/netdev/macsec.c [hunk 2]
```diff
@@ -0,0 +1,915 @@
...
+                r = sd_netlink_message_append_u32(m, MACSEC_SA_ATTR_PN, a->pn);
+                if (r < 0)
+                        return log_netdev_error_errno(netdev, r, "Could not append MACSEC_SA_ATTR_PN attribute: %m");
+        }
+
+        if (a->key_len) {
```
Comment (yuwata) [inline src/network/netdev/macsec.c]: `if (a->key_len > 0)`

File: src/network/netdev/macsec.c [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+        assert(netdev);
+        assert(netdev->state != _NETDEV_STATE_INVALID);
+
+        t = MACSEC(netdev);
+
+        assert(t);
+
+        r = sd_netlink_message_get_errno(m);
+        if (r == -EEXIST)
+                log_netdev_info(netdev, "netdev exists, using existing without changing its parameters");
```
Comment (yuwata) [inline src/network/netdev/macsec.c]: The log message seems to be invalid...

File: src/network/netdev/macsec.c [hunk 4]
```diff
@@ -0,0 +1,915 @@
...
+        }
+
+        log_netdev_debug(netdev, "MACsec add receive association configuration success");
+
+        return 1;
+}
+
+static int netdev_macsec_receive_association_fill_message(NetDev *netdev, ReceiveAssociation *a, sd_netlink_message **ret) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        MACsec *t;
```
Comment (yuwata) [inline src/network/netdev/macsec.c:328]: `t` seems not to be used.

File: src/network/netdev/macsec.c [hunk 5]
```diff
@@ -0,0 +1,915 @@
...
+
+        r = sd_netlink_message_close_container(m);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_RXSC_CONFIG attribute: %m");
+
+        return 0;
+}
+
+static int netdev_macsec_fill_message_sa(NetDev *netdev, TransmitAssociation *a, sd_netlink_message *m) {
+        MACsec *t;
```
Comment (yuwata) [inline src/network/netdev/macsec.c:258]: `t` seems not to be used.

File: src/network/netdev/macsec.c [hunk 6]
```diff
@@ -0,0 +1,915 @@
...
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_IFINDEX attribute: %m");
+
+        *ret = TAKE_PTR(m);
+
+        return 0;
+}
+
+static int netdev_macsec_fill_message_sci(NetDev *netdev, ReceiveChannel *c, sd_netlink_message *m) {
+        uint64_t sci;
+        MACsec *t;
```
Comment (yuwata) [inline src/network/netdev/macsec.c:226]: `t` seems not to be used.

File: src/network/netdev/macsec.c [hunk 7]
```diff
@@ -0,0 +1,915 @@
...
+        assert(netdev);
+        assert(m);
+
+        v = MACSEC(netdev);
+
+        r = sd_netlink_message_append_u16(m, IFLA_MACSEC_PORT, v->port);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append IFLA_MACSEC_PORT attribute: %m");
+
+        if (v->encrypt != -1) {
```
Comment (yuwata) [inline src/network/netdev/macsec.c]: `if (v->encrypt >= 0)`

File: src/network/netdev/macsec.c [hunk 8]
```diff
@@ -0,0 +1,915 @@
...
+        if (r < 0)
+                return r;
+
+        if (b)
+                r = parse_ip_port(rvalue, &b->rx.port);
+        else
+                r = parse_ip_port(rvalue, &c->port);
+
+        if (r < 0)
+                return 0;
```
Comment (yuwata) [inline src/network/netdev/macsec.c:565]: Please add log on failure.

File: src/network/netdev/macsec.c [hunk 9]
```diff
@@ -0,0 +1,915 @@
...
+        if (r < 0)
+                return r;
+
+        if (b)
+                r = config_parse_hwaddr(unit, filename, line, section, section_line, lvalue, ltype, rvalue, &b->rx.mac, userdata);
+        else
+                r = config_parse_hwaddr(unit, filename, line, section, section_line, lvalue, ltype, rvalue, &c->mac, userdata);
+
+        if (r < 0)
+                return 0;
```
Comment (yuwata) [inline src/network/netdev/macsec.c:613]: No log message.

File: src/network/netdev/netdev-gperf.gperf [hunk 10]
```diff
@@ -34,157 +35,169 @@ struct ConfigPerfItem;
...
+GENEVE.UDP6ZeroCheckSumRx,              config_parse_bool,                    0,                             offsetof(Geneve, udp6zerocsumrx)
+GENEVE.UDP6ZeroChecksumRx,              config_parse_bool,                    0,                             offsetof(Geneve, udp6zerocsumrx)
+GENEVE.UDP6ZeroCheckSumTx,              config_parse_bool,                    0,                             offsetof(Geneve, udp6zerocsumtx)
+GENEVE.UDP6ZeroChecksumTx,              config_parse_bool,                    0,                             offsetof(Geneve, udp6zerocsumtx)
+GENEVE.DestinationPort,                 config_parse_ip_port,                 0,                             offsetof(Geneve, dest_port)
+GENEVE.FlowLabel,                       config_parse_geneve_flow_label,       0,                             0
+MACSEC.Port,                            config_parse_ip_port,                 0,                             offsetof(MACsec, port)
+MACSEC.Encrypt,                         config_parse_bool,                    0,                             offsetof(MACsec, encrypt)
```
Comment (yuwata) [inline src/network/netdev/netdev-gperf.gperf]: Please use `config_parse_tristate()`.

**Comment by ssahani** [comment]
Updated thanks for the review @yuwata

File: man/systemd.netdev.xml [hunk 11]
```diff
@@ -851,6 +854,120 @@
...
+    <title>[MACSEC] Section Options</title>
+    <para>The <literal>[MACSEC]</literal> section only applies for
+    netdevs of kind <literal>macsec</literal>, and accepts the
+    following keys:</para>
+
+    <variablelist class='network-directives'>
+      <varlistentry>
+        <term><varname>Port=</varname></term>
+        <listitem>
+          <para>Specifies the to be used for the MACsec. Takes either value between 1 and 65535. This option is compulsory.</para>
```
Comment (topimiettinen) [inline man/systemd.netdev.xml]: Specifies the what?

File: src/libsystemd/sd-netlink/netlink-types.c [hunk 12]
```diff
@@ -16,6 +16,7 @@
 #include <linux/if_link.h>
 #include <linux/if_tunnel.h>
 #include <linux/l2tp.h>
+#include <linux/if_macsec.h>
```
Comment (topimiettinen) [inline src/libsystemd/sd-netlink/netlink-types.c]: Keeping the include list alphabetically sorted would need this to move upwards.

File: src/network/netdev/netdev.c [hunk 13]
```diff
@@ -15,6 +15,7 @@
 #include "netdev/ipvlan.h"
 #include "netdev/l2tp-tunnel.h"
 #include "netdev/macvlan.h"
+#include "netdev/macsec.h"
```
Comment (topimiettinen) [inline src/network/netdev/netdev.c:17]: Sorting order also here.

File: src/basic/parse-util.c [hunk 14]
```diff
@@ -566,6 +566,36 @@ int safe_atod(const char *s, double *ret_d) {
         return 0;
 }
 
+int safe_hexstring_aton(const char *s, uint8_t *b, unsigned int buf_len, unsigned int *len) {
+        unsigned int c = 0;
```
Comment (topimiettinen) [inline src/basic/parse-util.c]: I think c=0 is not necessary as it is always initialized later.

File: src/basic/parse-util.c [hunk 15]
```diff
@@ -566,6 +566,36 @@ int safe_atod(const char *s, double *ret_d) {
         return 0;
 }
 
+int safe_hexstring_aton(const char *s, uint8_t *b, unsigned int buf_len, unsigned int *len) {
```
Comment (topimiettinen) [inline src/basic/parse-util.c]: I'd use size_t for buf_len and *len.

File: man/systemd.netdev.xml [hunk 16]
```diff
@@ -851,6 +854,120 @@
...
+        <listitem>
+          <para>Specifies the packet number to be used for replay protection and the construction of
+          the initialization vector (along with the secure channel identifier [SCI]). Ranges a number between 1 and 4,294,967,295.
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>KeyId=</varname></term>
+        <listitem>
+          <para>Specifies the identification for they key. Ranges a number between 0 to 255 This option is compulsory.</para>
```
Comment (topimiettinen) [inline man/systemd.netdev.xml]: Period missing after 255.

File: man/systemd.netdev.xml [hunk 17]
```diff
@@ -851,6 +854,120 @@
...
+        </listitem>
+      </varlistentry>
+        <varlistentry>
+        <term><varname>Key=</varname></term>
+        <listitem>
+          <para>Specifies the identification for the key used in the transmit channel. That same key must be configured on the peer’s matching receive channel.
+          This option is compulsory. Takes a 128-bits key for example "dffafc8d7b9a43d5b9a3dfbbf6a30c16".</para>
```
Comment (topimiettinen) [inline man/systemd.netdev.xml]: I'd add comma after "key".

File: src/basic/parse-util.c [hunk 18]
```diff
@@ -566,6 +566,36 @@ int safe_atod(const char *s, double *ret_d) {
...
+
+        for(c = 0; c < buf_len && strlen(s) > 1; s += 2) {
+                unsigned int i;
+                char t[3] = {};
+
+                strncpy(t, s, 2);
+
+                errno = 0;
+                i = strtoul(t, &p, 16);
+                if (errno > 0 || i > 0xFF || *p)
```
Comment (topimiettinen) [inline src/basic/parse-util.c]: I'd use errno != 0.

File: src/basic/parse-util.c [hunk 19]
```diff
@@ -566,6 +566,36 @@ int safe_atod(const char *s, double *ret_d) {
...
+        assert(len);
+
+        if (strlen(s) % 2)
+                return -EINVAL;
+
+        for(c = 0; c < buf_len && strlen(s) > 1; s += 2) {
+                unsigned int i;
+                char t[3] = {};
+
+                strncpy(t, s, 2);
```
Comment (topimiettinen) [inline src/basic/parse-util.c]: Paraphrasing strncpy manual page: if there is no null byte among the first 2 bytes of s, the string placed in t will not be null-terminated. Please add t[2] = '\0'.

File: man/systemd.netdev.xml [hunk 20]
```diff
@@ -851,6 +854,120 @@
       </varlistentry>
     </variablelist>
   </refsect1>
+  <refsect1>
+    <title>[MACSEC] Section Options</title>
```
Comment (topimiettinen) [inline man/systemd.netdev.xml]: Shouldn't this be MACSec?

File: man/systemd.netdev.xml [hunk 21]
```diff
@@ -151,6 +151,9 @@
           <row><entry><varname>l2tp</varname></entry>
           <entry>A Layer 2 Tunneling Protocol (L2TP) is a tunneling protocol used to support virtual private networks (VPNs) or as part of the delivery of services by ISPs. It does not provide any encryption or confidentiality by itself</entry></row>
 
+          <row><entry><varname>macsec</varname></entry>
+          <entry>Media Access Control Security (MACsec) is an 802.1AE IEEE industry-standard security technology that provides secure communication for all traffic on Ethernet links. MACsec provides point-to-point security on Ethernet links between directly connected nodes and is capable of identifying and preventing most security threats.</entry></row>
```
Comment (topimiettinen) [inline man/systemd.netdev.xml:155]: MACSec?

**Comment by topimiettinen** [comment]
Keys could be seen as secrets that should not be accessible to unprivileged users, so later it would be nice to add a way to use a key file (preferably in such a format that it can be shared with all hosts in the network). The administrator could chmod the netdev file to 0700, but the contents would still be visible to unprivileged users with `systemctl show`.

**Comment by topimiettinen** [comment]
Please also wrap the long lines to make it easier to review in GitHub.

File: man/systemd.netdev.xml (hunk 21 above)
Comment (ssahani) [reply man/systemd.netdev.xml:155]: No it's called MACsec see https://www.linux.org/docs/man8/ip-macsec.html

File: man/systemd.netdev.xml (hunk 20 above)
Comment (ssahani) [reply man/systemd.netdev.xml]: No May be MACsec

File: src/basic/parse-util.c (hunk 14 above)
Comment (ssahani) [reply src/basic/parse-util.c]: yes right

File: src/basic/parse-util.c (hunk 15 above)
Comment (ssahani) [reply src/basic/parse-util.c]: why ? key_id and key is uint8_t type.

File: src/basic/parse-util.c (hunk 18 above)
Comment (ssahani) [reply src/basic/parse-util.c]: we use ```errno > 0``` .

File: src/basic/parse-util.c (hunk 19 above)
Comment (ssahani) [reply src/basic/parse-util.c]: would that be necessary it's initialized to `'\0'` any copying 2 bytes only or I am wrong .

File: src/libsystemd/sd-netlink/netlink-types.c (hunk 12 above)
Comment (ssahani) [reply src/libsystemd/sd-netlink/netlink-types.c]: yes thanks

**Comment by ssahani** [comment]
@poettering Please take a look.

**Comment by ssahani** [comment]
> Keys could be seen as secrets that should not be accessible to unprivileged users, so later it would be nice to add a way to use a key file (preferably in such a format that it can be shared with all hosts in the network). The administrator could chmod the netdev file to 0700, but the contents would still be visible to unprivileged users with `systemctl show`.

yes taken.

File: man/systemd.netdev.xml [hunk 22]
```diff
@@ -851,6 +854,120 @@
       </varlistentry>
     </variablelist>
   </refsect1>
+  <refsect1>
+    <title>[MACSEC] Section Options</title>
+    <para>The <literal>[MACSEC]</literal> section only applies for
+    netdevs of kind <literal>macsec</literal>, and accepts the
```
Comment (poettering) [inline man/systemd.netdev.xml]: "netdevs" → "network devices"

File: man/systemd.netdev.xml (hunk 11 above)
Comment (poettering) [inline man/systemd.netdev.xml]: "the to be used" → "the port to be used"?

Comment (poettering) [reply man/systemd.netdev.xml]: "either" shouldn't be there, should it?

File: man/systemd.netdev.xml [hunk 23]
```diff
@@ -851,6 +854,120 @@
...
+        <listitem>
+          <para>Takes a boolean. When true, enable encryption. Defaults to unset.</para>
+        </listitem>
+      </varlistentry>
+    </variablelist>
+  </refsect1>
+  <refsect1>
+    <title>[MACsecReceiveChannel] Section Options</title>
+    <para>The <literal>[MACsecReceiveChannel]</literal> section only applies for
+    netdevs of kind <literal>macsec</literal>, and accepts the
```
Comment (poettering) [inline man/systemd.netdev.xml]: netdevs → network devices

File: man/systemd.netdev.xml [hunk 24]
```diff
@@ -851,6 +854,120 @@
...
+    following keys:</para>
+
+    <variablelist class='network-directives'>
+      <varlistentry>
+        <term><varname>Port=</varname></term>
+        <listitem>
+          <para>Specifies the port to be used for the MACsec Receive Channel. The port used to make Secure Channel Identifier (SCI).
+          Takes either value between 1 and 65535. This option is compulsory.</para>
```
Comment (poettering) [inline man/systemd.netdev.xml]: either → a

File: man/systemd.netdev.xml [hunk 25]
```diff
@@ -851,6 +854,120 @@
...
+          <para>Specifies the MAC address to be used for the MACsec Receive Channel. The port used to make Secure Channel Identifier (SCI).
+          This option is compulsory.</para>
+        </listitem>
+      </varlistentry>
+    </variablelist>
+  </refsect1>
+   <refsect1>
+    <title>[MACsecTransmitAssociation] Section Options</title>
+    <para>The <literal>[MACsecTransmitAssociation]</literal> section only applies for
+    netdevs of kind <literal>macsec</literal>, and accepts the
```
Comment (poettering) [inline man/systemd.netdev.xml]: netdevs → network devices

File: man/systemd.netdev.xml (hunk 16 above)
Comment (poettering) [reply man/systemd.netdev.xml]: they key → the key

Comment (poettering) [reply man/systemd.netdev.xml]: Ranges a number between → Expects a number in the range.

File: man/systemd.netdev.xml [hunk 26]
```diff
@@ -851,6 +854,120 @@
...
+        <listitem>
+          <para>Specifies the packet number to be used for replay protection and the construction of
+          the initialization vector (along with the secure channel identifier [SCI]). Ranges a number between 1 and 4,294,967,295.
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>KeyId=</varname></term>
+        <listitem>
+          <para>Specifies the identification for they key. Ranges a number between 0 to 255. This option is compulsory.</para>
```
Comment (poettering) [inline man/systemd.netdev.xml]: they key → the key

"Ranges a number between" → "Expects a number in the range"

File: man/systemd.netdev.xml [hunk 27]
```diff
@@ -851,6 +854,120 @@
...
+    <para>The <literal>[MACsecTransmitAssociation]</literal> section only applies for
+    netdevs of kind <literal>macsec</literal>, and accepts the
+    following keys:</para>
+
+    <variablelist class='network-directives'>
+      <varlistentry>
+        <term><varname>PacketNumber=</varname></term>
+        <listitem>
+          <para>Specifies the packet number to be used for replay protection and the construction of
+          the initialization vector (along with the secure channel identifier [SCI]). Ranges a number between 1 and 4,294,967,295.
```
Comment (poettering) [inline man/systemd.netdev.xml]: "Ranges a number between" → "Expects a number between"

File: man/systemd.netdev.xml [hunk 28]
```diff
@@ -851,6 +854,120 @@
...
+        <term><varname>Port=</varname></term>
+        <listitem>
+          <para>Specifies the port to be used for the MACsec Receive Channel. The port used to make Secure Channel Identifier (SCI).
+          Takes either value between 1 and 65535. This option is compulsory.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>MACAddress=</varname></term>
+        <listitem>
+          <para>Specifies the MAC address to be used for the MACsec Receive Channel. The port used to make Secure Channel Identifier (SCI).
```
Comment (poettering) [inline man/systemd.netdev.xml]: "the port"? we are talking about  a MAC address here, no?

File: man/systemd.netdev.xml [hunk 29]
```diff
@@ -851,6 +854,120 @@
...
+      <varlistentry>
+        <term><varname>KeyId=</varname></term>
+        <listitem>
+          <para>Specifies the identification for they key. Ranges a number between 0 to 255. This option is compulsory.</para>
+        </listitem>
+      </varlistentry>
+        <varlistentry>
+        <term><varname>Key=</varname></term>
+        <listitem>
+          <para>Specifies the identification for the key used in the transmit channel. That same key must be configured on the peer’s matching receive channel.
```
Comment (poettering) [inline man/systemd.netdev.xml]: transmit → transmission?

File: man/systemd.netdev.xml [hunk 30]
```diff
@@ -851,6 +854,120 @@
...
+          This option is compulsory. Takes a 128-bits key, for example "dffafc8d7b9a43d5b9a3dfbbf6a30c16".</para>
```
Comment (poettering) [inline man/systemd.netdev.xml]: "Takes a 128-bit key encoded in a hexadecimal string, for example …"

File: man/systemd.netdev.xml [hunk 31]
```diff
@@ -851,6 +854,120 @@
...
+        </listitem>
+      </varlistentry>
+    </variablelist>
+   </refsect1>
+   <refsect1>
+    <title>[MACsecReceiveAssociation] Section Options</title>
+    <para>The <literal>[MACsecReceiveAssociation]</literal> section only applies for
+    netdevs of kind <literal>macsec</literal>, and accepts the
```
Comment (poettering) [inline man/systemd.netdev.xml]: network devices

File: src/basic/parse-util.c [hunk 32]
```diff
@@ -566,6 +566,36 @@ int safe_atod(const char *s, double *ret_d) {
         return 0;
 }
 
+int safe_hexstring_aton(const char *s, uint8_t *b, unsigned buf_len, unsigned int  *len) {
```
Comment (poettering) [inline src/basic/parse-util.c]: We already have unhexmem() which does just this, no?

File: src/network/netdev/macsec.c [hunk 33]
```diff
@@ -0,0 +1,911 @@
...
+        MACsec *t;
+        int r;
+
+        assert(netdev);
+        assert(m);
+
+        t = MACSEC(netdev);
+
+        assert(t);
+        assert(m);
```
Comment (poettering) [inline src/network/netdev/macsec.c]: you assert on "m" twice here

File: src/network/netdev/macsec.c [hunk 34]
```diff
@@ -0,0 +1,911 @@
...
+        assert(c);
+        assert(c->mac);
+        assert(c->port > 0);
+
+        r = sd_netlink_message_open_container(m, MACSEC_ATTR_RXSC_CONFIG);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_RXSC_CONFIG attribute: %m");
+
+        memcpy(&sci, c->mac, ETH_ALEN);
+        memcpy(((char *) &sci) + ETH_ALEN, &c->port, sizeof(c->port));
```
Comment (poettering) [inline src/network/netdev/macsec.c]: hmm, is this endian safe? I presume this should be handled different on different endianess, no?

Also, wouldn't it be nicer to solve this with a uinion? i.e.

```
union {
        uint64_t as_uint64;
        struct {
                struct ether_addr addr;
                uint16_t port;
        } as_address_and_port;
} sci;
```

Or something like that?

File: src/network/netdev/macsec.c [hunk 35]
```diff
@@ -0,0 +1,911 @@
...
+        else
+                r = macsec_receive_association_new_static(s, filename, section_line, &b);
+
+        if (r < 0)
+                return r;
+
+        if (b)
+                r = parse_ip_port(rvalue, &b->rx.port);
+        else
+                r = parse_ip_port(rvalue, &c->port);
```
Comment (poettering) [inline src/network/netdev/macsec.c:560]: is this actually an IP port? i.e. is 0 accepted here? parse_ip_port() filters that out... I mean, as I understand this is layer below IP, hence using an "IP port" parser is a bit of a layer violation, so it at least deserves a comment here, but I wonder what about port 0 here, does it have the same special meaning as on IP?

File: src/network/netdev/macsec.c [hunk 36]
```diff
@@ -0,0 +1,911 @@
...
+
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0,
```
Comment (poettering) [inline src/network/netdev/macsec.c]: why not parse `r` here to log_syntax(), why 0?

File: src/network/netdev/macsec.c [hunk 37]
```diff
@@ -0,0 +1,911 @@
...
+        if (r < 0)
+                return r;
+
+        if (b)
+                r = config_parse_hwaddr(unit, filename, line, section, section_line, lvalue, ltype, rvalue, &b->rx.mac, userdata);
+        else
+                r = config_parse_hwaddr(unit, filename, line, section, section_line, lvalue, ltype, rvalue, &c->mac, userdata);
+
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0,
```
Comment (poettering) [inline src/network/netdev/macsec.c]: 0 → r

File: src/network/netdev/macsec.c [hunk 38]
```diff
@@ -0,0 +1,911 @@
...
+        if (r < 0)
+                return r;
+
+        if (a)
+                r = safe_atou32(rvalue, &a->pn);
+        else
+                r = safe_atou32(rvalue, &b->sa.pn);
+
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0,
```
Comment (poettering) [inline src/network/netdev/macsec.c]: 0 → r

File: src/network/netdev/macsec.c [hunk 39]
```diff
@@ -0,0 +1,911 @@
...
+        else
+                r = macsec_receive_association_new_static(s, filename, section_line, &b);
+
+        if (r < 0)
+                return r;
+
+        if (a)
+                r = safe_atou32(rvalue, &a->pn);
+        else
+                r = safe_atou32(rvalue, &b->sa.pn);
```
Comment (poettering) [inline src/network/netdev/macsec.c]: according to your man page the value 0 is not OK here, needs to be refused

File: src/network/netdev/macsec.c [hunk 40]
```diff
@@ -0,0 +1,911 @@
...
+        if (r < 0)
+                return r;
+
+        if (a)
+                r = safe_hexstring_aton(rvalue, a->key, MACSEC_MAX_KEY_LEN, &a->key_len);
+        else
+                r = safe_hexstring_aton(rvalue, b->sa.key, MACSEC_MAX_KEY_LEN, &b->sa.key_len);
+
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Failed to parse key. Ignoring assignment: %s", rvalue);
```
Comment (poettering) [inline src/network/netdev/macsec.c]: 0 → r

File: src/network/netdev/macsec.c [hunk 41]
```diff
@@ -0,0 +1,911 @@
...
+        if (r < 0)
+                return r;
+
+        if (a)
+                r = safe_hexstring_aton(rvalue, a->key_id, MACSEC_KEYID_LEN, &len);
+        else
+                r = safe_hexstring_aton(rvalue, b->sa.key_id, MACSEC_KEYID_LEN, &len);
+
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Failed to parse key id. Ignoring assignment: %s", rvalue);
```
Comment (poettering) [inline src/network/netdev/macsec.c]: 0 → r

File: src/network/netdev/macsec.c [hunk 42]
```diff
@@ -0,0 +1,911 @@
...
+        if (streq(section, "MACsecTransmitAssociation"))
+                r = macsec_transmit_transmit_association_new_static(s, filename, section_line, &a);
+        else
+                r = macsec_receive_association_new_static(s, filename, section_line, &b);
+
+        if (r < 0)
+                return r;
+
+        if (a)
+                r = safe_hexstring_aton(rvalue, a->key_id, MACSEC_KEYID_LEN, &len);
```
Comment (poettering) [inline src/network/netdev/macsec.c]: hmm, you retrieve the size here but then don't use it

File: src/network/netdev/macsec.h [hunk 43]
```diff
@@ -0,0 +1,60 @@
...
+#include "networkd-util.h"
+
+typedef struct MACsec MACsec;
+
+typedef struct ReceiveChannel {
+        MACsec *mac_sec;
+        NetworkConfigSection *section;
+
+        uint16_t port;
+        struct ether_addr *mac;
```
Comment (poettering) [inline src/network/netdev/macsec.h]: hmm, why is this a ptr and not a buffer inline? what am I missing?

File: src/network/netdev/macsec.c (hunk 34 above)
Comment (ssahani) [reply src/network/netdev/macsec.c]: I tried to use it like 
```
memcpy(&sci.as_address_and_port.addr, &c->mac, ETH_ALEN);                                                                                                                          
memcpy(&sci.as_address_and_port.port, &c->port, sizeof(c->port));
```

Kernel does not accept it 

```
macsec-test: netdev could not be add receive association configurations: No such device

```

Kernel uses the same use. But we can keep union format.
https://elixir.bootlin.com/linux/latest/source/drivers/net/macsec.c#L412

File: src/network/netdev/macsec.c (hunk 35 above)
Comment (ssahani) [reply src/network/netdev/macsec.c:560]: The port used to make Secure Channel Identifier (SCI). ```0``` is a invalid value.

File: src/basic/parse-util.c (hunk 15 above)
Comment (ssahani) [reply src/basic/parse-util.c]: This function is dropped now.

File: man/systemd.netdev.xml (hunk 20 above)
Comment (yuwata) [reply man/systemd.netdev.xml]: +1 for MACsec.

File: src/network/netdev/macsec.c [hunk 44]
```diff
@@ -0,0 +1,928 @@
...
+}
+
+static int netdev_macsec_fill_message_sci(NetDev *netdev, ReceiveChannel *c, sd_netlink_message *m) {
+        union {
+                uint64_t as_uint64;
+
+                struct {
+                        struct ether_addr addr;
+                        uint16_t port;
+                } as_address_and_port;
```
Comment (yuwata) [inline src/network/netdev/macsec.c:224]: Not sure, but maybe `_packed_` attribute should be added to the struct for safety?

File: src/network/netdev/macsec.c [hunk 45]
```diff
@@ -0,0 +1,928 @@
...
+
+        assert(c);
+        assert(c->port > 0);
+
+        r = sd_netlink_message_open_container(m, MACSEC_ATTR_RXSC_CONFIG);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_RXSC_CONFIG attribute: %m");
+
+        memcpy(&sci.as_uint64, &c->mac, ETH_ALEN);
+        memcpy(((char *) &sci.as_uint64) + ETH_ALEN, &c->port, sizeof(c->port));
```
Comment (yuwata) [inline src/network/netdev/macsec.c]: Then now, this can be just a normal assignment: `sci.as_address_and_port.port = c->port;`

**Comment by yuwata** [comment]
@ssahani So, as usual, can I continue your work? Or are you still working on this? I'd like to implement `KeyFile=` to hide key from .netdev files, as similar to `WireGuard.PrivateKeyFile=`. And I'd like to add testcases for this.

Anyway, I will start to test this PR tomorrow, about 10h later :-)

File: src/network/netdev/macsec.c (hunk 44 above)
Comment (ssahani) [reply src/network/netdev/macsec.c:224]: I tried that. Did not worked. I mean kernel did not accepted the value. But just for sake can be.

**Comment by ssahani** [comment]
If you interested please do it. I like this mixed contribution. Feels like a team.

File: src/network/netdev/macsec.c (hunk 45 above)
Comment (ssahani) [reply src/network/netdev/macsec.c]: I have not tested this. Looking at the kernel code they do this in same manner. It also validates that.

File: man/systemd.netdev.xml (hunk 20 above)
Comment (ssahani) [reply man/systemd.netdev.xml]: Done

File: src/network/netdev/macsec.c (hunk 45 above)
Comment (ssahani) [reply src/network/netdev/macsec.c]: that works. replaced ;

**Comment by ssahani** [comment]
@yuwata I have addressed what were remaining. Please take it from here. thanks. I am going to work on tc now.

**Comment by yuwata** [comment]
OK, I will start to test this PR from this state.

**Comment by yuwata** [comment]
I've opened #12222, which contains a revised version of this PR. Let's close this.

**Pull request closed without merging**